records/2023-03-28.secret-management.md

```diff
@@ -0,0 +1,189 @@
+
+### AWS Parameter Store
+[AWS Parameter STore](https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-parameter-store.html) is a first-party AWS offering that integrates natively with AWS resources, and can be implemented within Kubernetes using a CSI Driver. It is built to manage parameters within AWS but is encrypted and could theoretically be used to manage secrets as well.
+
+## Features Comparison
+
+|                     | Fine Grained RBAC | Multiple Secret Stores | Multiple Secret  Backends | Automatic Secret Rotation | Just In Time Authentication | Cloud Agnostic | Managed Service | Private Connection Required | License Cost | People-Time Cost | Infrastructure Cost |
+|---------------------|-------------------|------------------------|---------------------------|---------------------------|-----------------------------|----------------|-----------------|--------------|--------------|------------------|---------------------|
+| HCP Vault           | Yes               | Yes                    | Yes                       | Yes                       | Yes                         | Yes            | Yes             | Yes          | ~ $20k/Year  | Low              | Included w/ License |
+| Self Hosted Vault   | Yes               | Yes                    | Yes                       | Yes                       | Yes                         | Yes            | No              | No           | Free         | High             | Medium              |
+| AWS Secrets Manager | Yes               | No (One per account)   | No                        | Yes                       | No                          | No             | Yes             | No           | Free         | Medium           | Low                 |
+| AWS Parameter Store | Yes               | No (One per account)   | No                        | No                        | No                          | No             | Yes             | No           | Free         | Medium           | Free                |
+
+## Additional considerations
+
+### Vault - Common
+
+#### Kubernetes Integration
+
+[Kubernetes Integration is done via the vault secret injector combined with the vault sidecar.](https://developer.hashicorp.com/vault/tutorials/kubernetes/kubernetes-external-vault)
+
+The secret injector is scoped to a service account within kubernetes, and creates a JWT token that is used to authenticate with Vault. The sidecar is a container that runs in parallel with the target pod (e.g. admin or api). The sidecar periodically checks Vault for updates to the secrets and automatically injects the secrets into the target pod. When the secret is updated, the pod does not need to be refreshed, however the code must be written in such a way that it will automatically check for new values for this to be effective. The integration will require modifications to the Kustomize (or Helm) deployment for each application that needs to use it. These modifications are referenced just once. 
```

Comment:
Currently we read the env vars on app start and store them in a `Config` class, probably we could just change this to make a getter for each that fetched a fresh copy from the env vars.

Reply:
Yeah we could start with this for our abstraction, and if we want and possible later on, actually repopulate on demand.

Reply:
Yeah this would be my instinct as well - putting on my old java developer hat - no idea if it's valid for python paradigms - having a config class is great, just have an init function in there that pulls from the OS environment variables, and a reload that triggers the config.